packages/data-broker/src/services/superset/SupersetService.js

```diff
@@ -0,0 +1,154 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+import moment from 'moment';
+import { Service } from '../Service';
+import { getSupersetApiInstance } from './getSupersetApi';
+
+export class SupersetService extends Service {
+  /**
+   * @param {ModelRegistry} models
+   */
+  constructor(models) {
+    super(models);
+    this.pullers = {
+      [this.dataSourceTypes.DATA_ELEMENT]: this.pullAnalytics.bind(this),
+      [this.dataSourceTypes.DATA_GROUP]: this.pullEvents.bind(this),
+      [this.dataSourceTypes.SYNC_GROUP]: this.pullSyncGroups.bind(this),
+    };
+  }
+
+  async push() {
+    throw new Error('Data push is not supported in SupersetService');
+  }
+
+  async delete() {
+    throw new Error('Data deletion is not supported in SupersetService');
+  }
+
+  async pull(dataSources, type, options) {
+    const puller = this.pullers[type];
+    return puller(dataSources, options);
+  }
+
+  async pullMetadata() {
+    throw new Error('pullMetadata is not supported in SupersetService');
+  }
+
+  /**
+   * @param {DataElement[]} dataSources
+   * @param {} options
+   * @private
+   */
+  async pullAnalytics(dataSources, options) {
+    let mergedResults = [];
+    for (const [supersetInstanceCode, instanceDataSources] of Object.entries(
+      this.groupBySupersetInstanceCode(dataSources),
+    )) {
+      const supersetInstance = await this.models.supersetInstance.findOne({
+        code: supersetInstanceCode,
+      });
+      if (!supersetInstance)
+        throw new Error(`No superset instance found with code "${supersetInstanceCode}"`);
+      const api = await getSupersetApiInstance(this.models, supersetInstance);
+      for (const [chartId, chartDataSources] of Object.entries(
+        this.groupByChartId(instanceDataSources),
+      )) {
+        const results = await this.pullForApiForChart(
+          api,
+          chartId,
+          chartDataSources.map(de => de.code),
+        );
+        mergedResults = mergedResults.concat(results);
+      }
+    }
+
+    return {
+      results: mergedResults,
+      // TODO: either implement properly in #NOT-521 or remove entirely in #NOT-522
+      metadata: { dataElementCodeToName: {} },
+    };
+  }
+
+  /**
+   * @param {SupersetApi} api
+   * @param {string} chartId
+   * @param {string[]} dataElementCodes
+   * @return {Promise<Object[]>} analytic results
+   * @private
+   */
+  async pullForApiForChart(api, chartId, dataElementCodes) {
+    const response = await api.chartData(chartId);
+    const { data } = response.result[0];
+
+    const results = [];
+    for (const datum of data) {
+      const universalCode = datum.item_code;
+      const dataElementCode = dataElementCodes.find(code => code.includes(universalCode));
```

Comment:
It was a convention, but yeah you're right it's a bit bad hardcoding it in. I've added a config option supersetItemCode which we can add to data elements for this mapping. Otherwise if not found it defaults to matching the tupaia data element code with the superset item code.

Reply:
Yup, nice, that sounds good 👍